### PR TITLE
Modifying and deleting candidates

### PR DESCRIPTION
**Pull request by Elrhino**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -14,7 +14,7 @@
 
         <!-- client -->
         <gwt.version>2.7.0</gwt.version>
-        <gwtp.version>1.4</gwtp.version>
+        <gwtp.version>1.5</gwtp.version>
```
Comment (olafleur) [inline pom.xml]: why not 1.5.3 ? ;)

File: src/main/java/com/arcbees/bourseje/client/admin/edit/EditView.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    @Override
+    protected void onAttach() {
+        uploader = new SingleUploader(
+                IFileInput.FileInputType.CUSTOM.withZone(imagePlaceHolder).with(imagePlaceHolder),
+                new BaseUploadStatus(),
+                null);
+        uploader.setAutoSubmit(false);
+        uploader.setValidExtensions(".jpg, .gif, .png, .jpeg");
+
+        // handlers
```
Comment (olafleur) [inline src/main/java/com/arcbees/bourseje/client/admin/edit/EditView.java]: comment needed?

**Comment by olafleur** [comment]
LGTM

File: src/main/java/com/arcbees/bourseje/client/admin/dashboard/candidate/CandidateAdminView.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    @Override
+    public void setCandidate(Candidate candidate, int nbOfVotes) {
+        name.setInnerHTML(candidate.getName());
+        company.setInnerHTML(candidate.getCompany());
+        image.setSrc(candidate.getPicture());
+        nbVotes.setInnerHTML(String.valueOf(nbOfVotes));
+    }
+
+    @UiHandler("modify")
+    public void onModify(ClickEvent event) {
```
Comment (christiangoudreau) [inline src/main/java/com/arcbees/bourseje/client/admin/dashboard/candidate/CandidateAdminView.java]: No need for public keyword on @UiHandler method. They should not be public in fact.

File: src/main/java/com/arcbees/bourseje/client/admin/dashboard/candidate/CandidateAdminView.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        Candidate candidate = new Candidate(name.getInnerText(), company.getInnerText(), image.getSrc());
+
+        getUiHandlers().onModify(candidate);
+    }
+
+    @UiHandler("delete")
+    public void onDelete(ClickEvent event) {
```
Comment (christiangoudreau) [inline src/main/java/com/arcbees/bourseje/client/admin/dashboard/candidate/CandidateAdminView.java]: Idem and on all the classes you did that.

File: src/main/java/com/arcbees/bourseje/client/NameTokens.java [hunk 5]
```diff
@@ -35,4 +35,5 @@
     public static final String ADMIN_DASHBOARD = PREFIX_ADMIN + "/dashboard";
     public static final String WINNER = PREFIX_ADMIN + "/winner";
     public static final String ADD = PREFIX_ADMIN + "/add";
+    public static final String EDIT = PREFIX_ADMIN+ "/edit";
```
Comment (meriouma) [inline src/main/java/com/arcbees/bourseje/client/NameTokens.java]: Missing space before `+`

File: src/main/java/com/arcbees/bourseje/client/admin/edit/EditPresenter.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+import com.gwtplatform.mvp.client.annotations.NameToken;
+import com.gwtplatform.mvp.client.annotations.ProxyStandard;
+import com.gwtplatform.mvp.client.proxy.PlaceManager;
+import com.gwtplatform.mvp.client.proxy.ProxyPlace;
+import com.gwtplatform.mvp.shared.proxy.PlaceRequest;
+
+public class EditPresenter extends Presenter<EditPresenter.MyView, EditPresenter.MyProxy>
+        implements EditUiHandlers {
+
+    private Candidate candidateToUpdate;
```
Comment (meriouma) [inline src/main/java/com/arcbees/bourseje/client/admin/edit/EditPresenter.java]: field ordering

File: src/main/java/com/arcbees/bourseje/client/admin/edit/EditPresenter.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    @ProxyStandard
+    @NameToken(NameTokens.EDIT)
+    interface MyProxy extends ProxyPlace<EditPresenter> {
+    }
+
+    private final RestDispatch dispatch;
+    private final AdminService adminService;
+    private CandidateService candidateService;
```
Comment (meriouma) [inline src/main/java/com/arcbees/bourseje/client/admin/edit/EditPresenter.java]: can be final

File: src/main/java/com/arcbees/bourseje/client/admin/edit/EditPresenter.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+            public void onSuccess(Candidate candidate) {
+                candidateToUpdate = candidate;
+                getView().setCandidate(candidate);
+            }
+        });
+    }
+
+    @Override
+    public void onAddCandidateClicked(Candidate candidate) {
+        dispatch.execute(adminService.updateCandidate(candidateToUpdate.getName(), candidate), new AdminRestCallback<Void>() {
```
Comment (meriouma) [inline src/main/java/com/arcbees/bourseje/client/admin/edit/EditPresenter.java]: line length?

File: src/main/java/com/arcbees/bourseje/client/admin/edit/EditView.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        company.setValue(candidate.getCompany());
+        url = candidate.getPicture();
+        imagePlaceHolder.setImageSource(url);
+    }
+
+    private void initButtons() {
+        $(updateCandidate).on(CLICK, new Function() {
+            @Override
+            public void f() {
+                if (!name.getValue().isEmpty() && !company.getValue().isEmpty()) {
```
Comment (meriouma) [inline src/main/java/com/arcbees/bourseje/client/admin/edit/EditView.java]: Do you need to check the `url` is provided too?

File: src/main/java/com/arcbees/bourseje/server/services/AdminService.java [hunk 10]
```diff
@@ -59,4 +59,17 @@ private void clearVotesIfNotAlreadyStarted() {
     public void addCandidate(Candidate candidate) {
         candidateDao.put(candidate);
     }
+
+    public void updateCandidate(String name, Candidate candidate) {
+        Candidate oldCandidate = candidateDao.getByCandidateName(name);
+
+        candidateDao.delete(oldCandidate);
```
Comment (meriouma) [inline src/main/java/com/arcbees/bourseje/server/services/AdminService.java]: Not sure why you delete the old one, instead of merging the 2 candidates?

**Pull request merged**